File: app/src/main/java/closer/vlllage/com/closer/handler/PermissionHandler.java
package closer.vlllage.com.closer.handler;

import android.app.Activity;
import android.content.pm.PackageManager;
import android.os.Build;
import android.support.v4.app.ActivityCompat;

import java.util.HashSet;
import java.util.Set;

import closer.vlllage.com.closer.pool.PoolMember;
import io.reactivex.disposables.Disposable;
import io.reactivex.subjects.PublishSubject;

public class PermissionHandler extends PoolMember {

    private static final int REQUEST_CODE_PERMISSION = 1009293;
    private PublishSubject<String> permissionChanges = PublishSubject.create();

    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        if (requestCode != REQUEST_CODE_PERMISSION) {
            return;
        }

        for (String permission : permissions) {
            permissionChanges.onNext(permission);
        }
    }

    public LocationCheck check(String... permissions) {
        LocationCheck check;

        if (has(permissions)) {
            check = new LocationCheck(true);
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            check = new LocationCheck(permissions);
            $(ActivityHandler.class).getActivity().requestPermissions(permissions, REQUEST_CODE_PERMISSION);
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            check = new LocationCheck(permissions);
            ActivityCompat.requestPermissions($(ActivityHandler.class).getActivity(), permissions, REQUEST_CODE_PERMISSION);
        } else {
            check = new LocationCheck(has(permissions));
        }

        return check;
    }

    public boolean has(String... permissions) {
        Activity activity = $(ActivityHandler.class).getActivity();

        for (String permission : permissions) {
            if (ActivityCompat.checkSelfPermission(activity, permission) != PackageManager.PERMISSION_GRANTED) {
                return false;
            }
        }

        return true;
    }

    public class LocationCheck {

        private Boolean override;
        private String[] permissions;
        private final Set<String> permissionsGranted = new HashSet<>();
        private PermissionCallback callback;
        private Disposable disposable;

        LocationCheck(String[] permissions) {
            this.permissions = permissions;
        }

        private LocationCheck(boolean granted) {
            override = granted;
        }

        public void when(PermissionCallback callback) {
            if (override != null) {
                callback.onPermissionResult(override);
                return;
            }

            if (permissions == null || permissions.length == 0) {
                callback.onPermissionResult(true);
                return;
            }

            this.callback = callback;

            subscribe();
        }

        private void subscribe() {
            disposable = permissionChanges.filter(permission -> {
                for (String p : permissions) {
                    if (p.equals(permission)) {
                        return true;
                    }
                }

                return false;
            }).subscribe(
                    permission -> {
                        if (has(permission)) {
                            permissionsGranted.add(permission);
                            if (permissionsGranted.size() == permissions.length) {
                                callback.onPermissionResult(true);
                                $(DisposableHandler.class).dispose(disposable);
                            }
                        } else {
                            callback.onPermissionResult(false);
                            $(DisposableHandler.class).dispose(disposable);
                        }
                    }
            );

            $(DisposableHandler.class).add(disposable);
        }
    }

    public interface PermissionCallback {
        void onPermissionResult(boolean granted);
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/data/PermissionHandler.kt
package closer.vlllage.com.closer.handler.data

import android.content.pm.PackageManager
import androidx.core.app.ActivityCompat
import closer.vlllage.com.closer.handler.helpers.ActivityHandler
import closer.vlllage.com.closer.handler.helpers.ApplicationHandler
import closer.vlllage.com.closer.handler.helpers.DefaultAlerts
import closer.vlllage.com.closer.handler.helpers.DisposableHandler
import com.queatz.on.On
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.subjects.PublishSubject
import java.util.*

class PermissionHandler constructor(private val on: On) {
    companion object {
        private val permissionChanges = PublishSubject.create<String>()
        private const val REQUEST_CODE_PERMISSION = 1009293
    }

    fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode != REQUEST_CODE_PERMISSION) {
            return
        }

        permissions.forEach { permissionChanges.onNext(it) }
    }

    fun check(vararg permissions: String): LocationCheck {
        return if (has(*permissions)) {
            LocationCheck(true)
        } else {
            on<ActivityHandler>().activity?.let {
                it.requestPermissions(permissions, REQUEST_CODE_PERMISSION)
                LocationCheck(arrayOf(*permissions))
            } ?: LocationCheck(false)
        }
    }

    fun has(vararg permissions: String): Boolean {
        val context = on<ApplicationHandler>().app

        for (permission in permissions) {
            if (ActivityCompat.checkSelfPermission(context, permission) != PackageManager.PERMISSION_GRANTED) {
                return false
            }
        }

        return true
    }

    fun denied(vararg permissions: String): Boolean {
        val activity = on<ActivityHandler>().activity

        if (has(*permissions)) {
            return false
        }

        for (permission in permissions) {
            if (!ActivityCompat.shouldShowRequestPermissionRationale(activity!!, permission)) {
                return true
            }
        }

        return false
    }

    inner class LocationCheck {

        private var override: Boolean? = null
        private var permissions: Array<String> = arrayOf()
        private var permissionsGranted = HashSet<String>()
        private var callback: ((granted: Boolean) -> Unit)? = null
        private var disposable: Disposable? = null

        constructor(permissions: Array<String>) {
            this.permissions = permissions
        }

        constructor(granted: Boolean) {
            override = granted
        }

        fun `when`(callback: ((granted: Boolean) -> Unit)) {
            if (override != null) {
                callback.invoke(override!!)
                return
            }

            if (permissions.isEmpty()) {
                callback.invoke(true)
                return
            }

            this.callback = callback

            subscribe()
        }

        private fun subscribe() {
            disposable = permissionChanges.filter { permission ->
                for (p in permissions) {
                    if (p == permission) {
                        return@filter true
                    }
                }

                false
            }.subscribe(
                    { permission ->
                        if (has(permission)) {
                            permissionsGranted.add(permission)
                            if (permissionsGranted.size == permissions.size) {
                                callback!!.invoke(true)
                                on<DisposableHandler>().dispose(disposable!!)
                            }
                        } else {
                            callback!!.invoke(false)
                            on<DisposableHandler>().dispose(disposable!!)
                        }
                    }, { error -> on<DefaultAlerts>().thatDidntWork() })

            on<DisposableHandler>().add(disposable!!)
        }
    }
}
